core/payments/ondemand/dynamostore/dynamodb_cumulative_payment_store.go

```diff
@@ -0,0 +1,213 @@
+type DynamoDBCumulativePaymentStore struct {
+	// The DynamoDB client to use for storage operations
+	dynamoClient *dynamodb.Client
+	// The name of the DynamoDB table to store payments in, stored as *string for use in DynamoDB operations
+	tableName *string
+	// The account address, pre-built as a key for DynamoDB operations
+	accountKey map[string]types.AttributeValue
+}
+
+var _ ondemand.CumulativePaymentStore = (*DynamoDBCumulativePaymentStore)(nil)
+
+// Creates a new DynamoDB-backed cumulative payment store
+func NewDynamoDBCumulativePaymentStore(
+	dynamoClient *dynamodb.Client,
+	tableName string,
+	// The account ID this store is tracking payments for
+	accountID gethcommon.Address,
+) (*DynamoDBCumulativePaymentStore, error) {
+	if dynamoClient == nil {
+		return nil, fmt.Errorf("dynamoClient cannot be nil")
+	}
+	if tableName == "" {
+		return nil, fmt.Errorf("tableName cannot be empty")
+	}
+	if accountID == (gethcommon.Address{}) {
+		return nil, fmt.Errorf("accountID cannot be the zero address")
+	}
+
+	return &DynamoDBCumulativePaymentStore{
+		dynamoClient: dynamoClient,
+		tableName:    aws.String(tableName),
+		accountKey: map[string]types.AttributeValue{
+			attributeAccountID: &types.AttributeValueMemberS{Value: accountID.Hex()},
+		},
+	}, nil
+}
+
+// AddCumulativePayment atomically adds a given amount to the cumulative payment
+//
+// This method uses DynamoDB's conditional update feature to ensure thread-safe operations.
+// It performs an atomic read-modify-write operation that:
+// 1. Adds the amount to the existing cumulative payment (or initializes to the amount if no record exists)
+// 2. Ensures the resulting value doesn't exceed maxCumulativePayment
+//
+// Returns the new cumulative payment value after the addition.
+// Returns an InsufficientFundsError if the addition would exceed maxCumulativePayment.
+func (s *DynamoDBCumulativePaymentStore) AddCumulativePayment(
+	ctx context.Context,
+	amount *big.Int,
+	maxCumulativePayment *big.Int,
+) (*big.Int, error) {
+	if amount == nil {
+		return nil, errors.New("amount cannot be nil")
+	}
+	if amount.Sign() <= 0 {
+		return nil, fmt.Errorf("amount must be positive (decrementing not supported): received %s", amount.String())
+	}
+
+	if maxCumulativePayment == nil {
+		return nil, errors.New("maxCumulativePayment cannot be nil")
+	}
+	if maxCumulativePayment.Sign() <= 0 {
+		return nil, fmt.Errorf("maxCumulativePayment must be positive: received %s", maxCumulativePayment.String())
+	}
+
+	// Calculate the maximum allowed current cumulative payment value.
+	// If the cumulativePayment value that exists prior to the addition is greater than this, that means the account
+	// doesn't have enough funds to pay
+	maxAllowedCurrent := new(big.Int).Sub(maxCumulativePayment, amount)
+
+	// if maxAllowedCurrent is negative, that means the cost of this single dispersal alone exceeds total deposits
+	if maxAllowedCurrent.Sign() < 0 {
+		currentValue, err := s.getCurrentPayment(ctx)
```

Comment:
you're assuming no concurrent operation can modify the currentPayment. Is that accurate?